

<!DOCTYPE html>
<html class="writer-html5" lang="en">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Integrate three mouse brain sections in the RNA modality &mdash; spamosaic v1.0.3 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="../../../_static/nbsphinx-code-cells.css?v=2aa19091" />
      <link rel="stylesheet" type="text/css" href="../../../_static/custom.css?v=9af0dbfa" />

  
      <script src="../../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js?v=393fe109"></script>
      <script src="../../../_static/doctools.js?v=888ff710"></script>
      <script src="../../../_static/sphinx_highlight.js?v=4825356b"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
      <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
      <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="Integrate three tonsil sections in the RNA modality" href="Tonsil_RNA_horizontal.html" />
    <link rel="prev" title="Horizontal Integration" href="index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            spamosaic
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../../install.html">Installation</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../index.html">Tutorials</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="index.html">Horizontal Integration</a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="#">Integrate three mouse brain sections in the RNA modality</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#preprocessing">preprocessing</a></li>
<li class="toctree-l4"><a class="reference internal" href="#training">training</a></li>
<li class="toctree-l4"><a class="reference internal" href="#inference">inference</a></li>
<li class="toctree-l4"><a class="reference internal" href="#clustering">clustering</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Tonsil_RNA_horizontal.html">Integrate three tonsil sections in the RNA modality</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../vertical/index.html">Vertical Integration</a></li>
<li class="toctree-l2"><a class="reference internal" href="../mosaic/index.html">Mosaic Integration</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../imputation/index.html">Imputation</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../spamosaic.html">API Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../dataset.html">Datasets</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../credits.html">Credits</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">spamosaic</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">Tutorials</a></li>
          <li class="breadcrumb-item"><a href="index.html">Horizontal Integration</a></li>
      <li class="breadcrumb-item active">Integrate three mouse brain sections in the RNA modality</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../../_sources/tutorials/integration/horizontal/MB_RNA_horizontal.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="Integrate-three-mouse-brain-sections-in-the-RNA-modality">
<h1>Integrate three mouse brain sections in the RNA modality<a class="headerlink" href="#Integrate-three-mouse-brain-sections-in-the-RNA-modality" title="Permalink to this heading"></a></h1>
<p>In this notebook, we used SpaMosaic to integrate three postnatal mouse brain sections in the RNA modality. All three sections were measured by spatial-RNA-epigenome seq (<a class="reference external" href="https://www.nature.com/articles/s41586-023-05795-1">Zhang et al., Nature. 2023</a>). The first section consists of the transcriptome and ATAC profiles of a P21 mouse brain section in a capture area with 50×50 pixels, each 20 µm in diameter. The second section is a replicate experiment for the first section. It measured the
spatial transcriptome and ATAC of a P21 mouse brain section with a capture array of 50×50 pixels, each 20 µm in diameter. The third section consists of the transcriptome and histone modification (H3K27ac) assays of a P21 mouse brain section with a capture array of 50×50 pixels, each 20 µm in diameter. We retained the transcriptome profiles of three sections for horizontal settings.</p>
<p>Data used in this notebook can be downloaded from <a class="reference external" href="https://drive.google.com/drive/folders/1FM42Ibhrf5l6cjzORTFqEXLuxlT5ltJ2?usp=drive_link">google drive</a>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">from</span> <span class="nn">os.path</span> <span class="kn">import</span> <span class="n">join</span>

<span class="kn">import</span> <span class="nn">spamosaic</span>
<span class="kn">from</span> <span class="nn">spamosaic.framework</span> <span class="kn">import</span> <span class="n">SpaMosaic</span>

<span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="p">[</span><span class="s1">&#39;CUBLAS_WORKSPACE_CONFIG&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;:4096:8&#39;</span>  <span class="c1"># for CuBLAS operation and you have CUDA &gt;= 10.2</span>

<span class="kn">import</span> <span class="nn">spamosaic.utils</span> <span class="k">as</span> <span class="nn">utls</span>
<span class="kn">from</span> <span class="nn">spamosaic.preprocessing</span> <span class="kn">import</span> <span class="n">RNA_preprocess</span><span class="p">,</span> <span class="n">ADT_preprocess</span><span class="p">,</span> <span class="n">Epigenome_preprocess</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ad1_rna</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s1">&#39;./s1_adata_rna.h5ad&#39;</span><span class="p">)</span>
<span class="n">ad2_rna</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s1">&#39;./s2_adata_rna.h5ad&#39;</span><span class="p">)</span>
<span class="n">ad3_rna</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s1">&#39;./s3_adata_rna.h5ad&#39;</span><span class="p">)</span>

<span class="c1"># change the batch label; `src` is the key in the row metadata that holds the batch labels</span>
<span class="n">ad3_rna</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;src&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;s3&#39;</span>
<span class="c1"># modify the barcode</span>
<span class="n">ad3_rna</span><span class="o">.</span><span class="n">obs_names</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;s1-&#39;</span><span class="p">,</span> <span class="s1">&#39;s3-&#39;</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">ad3_rna</span><span class="o">.</span><span class="n">obs_names</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># intersection of features</span>
<span class="n">shared_genes</span> <span class="o">=</span> <span class="n">ad1_rna</span><span class="o">.</span><span class="n">var_names</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">ad2_rna</span><span class="o">.</span><span class="n">var_names</span><span class="p">)</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">ad3_rna</span><span class="o">.</span><span class="n">var_names</span><span class="p">)</span>
</pre></div>
</div>
</div>
<section id="preprocessing">
<h2>preprocessing<a class="headerlink" href="#preprocessing" title="Permalink to this heading"></a></h2>
<p>SpaMosaic expects batch-corrected low-dimensional representations for each modality as input. Therefore, for the raw count assays of each modality, SpaMosaic performs feature selection and dimension reduction to obtain the low-dimensional representations. Then, Harmony is performed to integrate these representations in each modality (optional, depending on the presence of strong batch effects).</p>
<p>By default, the feature selection and dimension reduction consist of following steps:</p>
<ul class="simple">
<li><p>RNA assays: highly variable gene selection <span class="math notranslate nohighlight">\(\rightarrow\)</span> log normalization <span class="math notranslate nohighlight">\(\rightarrow\)</span> PCA</p></li>
<li><p>epigenome assays: highly variable peak selection <span class="math notranslate nohighlight">\(\rightarrow\)</span> TF-IDF transformation <span class="math notranslate nohighlight">\(\rightarrow\)</span> log normalization <span class="math notranslate nohighlight">\(\rightarrow\)</span> PCA</p></li>
<li><p>protein assays: centered log ratio normalization <span class="math notranslate nohighlight">\(\rightarrow\)</span> PCA</p></li>
</ul>
<p>In horizontal integration, SpaMosaic requires the input in the format of</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="p">{</span>
    <span class="s1">&#39;modality_name&#39;</span><span class="p">:</span> <span class="p">[</span><span class="n">adata1</span><span class="p">,</span> <span class="n">adata2</span><span class="p">,</span> <span class="n">adata3</span><span class="p">,</span> <span class="o">...</span><span class="p">]</span>
<span class="p">}</span>
</pre></div>
</div>
<p>In the dictionary, each key represents a modality and each modality key corresponds to list of <code class="docutils literal notranslate"><span class="pre">anndata</span></code> objects. Each <code class="docutils literal notranslate"><span class="pre">anndata</span></code> object contains modality-specific information for a particular section.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">input_dict</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;rna&#39;</span><span class="p">:</span>  <span class="p">[</span><span class="n">ad1_rna</span><span class="p">[:,</span> <span class="n">shared_genes</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">(),</span>  <span class="n">ad2_rna</span><span class="p">[:,</span> <span class="n">shared_genes</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">(),</span> <span class="n">ad3_rna</span><span class="p">[:,</span> <span class="n">shared_genes</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()],</span>
<span class="p">}</span>

<span class="n">input_key</span> <span class="o">=</span> <span class="s1">&#39;dimred_bc&#39;</span>
</pre></div>
</div>
</div>
<p>Parameters in the following <code class="docutils literal notranslate"><span class="pre">RNA_preprocess</span></code> function:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">batch_corr</span></code>: whether to perform batch correction with Harmony</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">n_hvg</span></code>: how many highly variable genes to select</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">batch_key</span></code>: the key in the row metadata that holds the batch labels</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">key</span></code>: the key in <code class="docutils literal notranslate"><span class="pre">.obsm</span></code> that will hold the output low-dimensional representations</p></li>
</ul>
<p>After preprocessing, the low-dimensional representations can be accessed by: <code class="docutils literal notranslate"><span class="pre">ad1_rna.obsm[input_key]</span></code>, <code class="docutils literal notranslate"><span class="pre">ad2_rna.obsm[input_key]</span></code>, …</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">RNA_preprocess</span><span class="p">(</span><span class="n">input_dict</span><span class="p">[</span><span class="s1">&#39;rna&#39;</span><span class="p">],</span> <span class="n">batch_corr</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">n_hvg</span><span class="o">=</span><span class="mi">5000</span><span class="p">,</span> <span class="n">batch_key</span><span class="o">=</span><span class="s1">&#39;src&#39;</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="n">input_key</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Use GPU mode.
        Initialization is completed.
        Completed 1 / 10 iteration(s).
        Completed 2 / 10 iteration(s).
        Completed 3 / 10 iteration(s).
Reach convergence after 3 iteration(s).
</pre></div></div>
</div>
</section>
<section id="training">
<h2>training<a class="headerlink" href="#training" title="Permalink to this heading"></a></h2>
<p>SpaMosaic employs modality-specific graph neural networks to embed each modality’s input into latent space. In horizontal integration, all sections have only one modality, thus each section has only one set of embeddings.</p>
<p>The crucial parameters include:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">intra_knns</span></code>: how many nearest neighbors to consider when searching for spatial neighbors within each section (list or integer)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">inter_knn_base</span></code>: how many nearest neighbors to consider when searching for mutual nearest neighbors between sections (integer)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">w_g</span></code>: the weight for spatial-adjacency graph</p></li>
</ul>
<p>The following parameters are recommended to use in complex integration scenarios, like varying resolution or size across sections</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">smooth_input</span></code>: whethere to smooth the input representations (bool)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">smooth_L</span></code>: number of LGCN layers for smoothing input</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">inter_auto_knn</span></code>: whether to automatically balance the kNN during MNN searching (bool)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">rmv_outlier</span></code>: whether to remove outlier of MNN (bool)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">contamination</span></code>: percentage of removed MNN outlier</p></li>
</ul>
<p>for training:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">net</span></code>: which graph neural network to use (only support wlgcn now)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">lr</span></code>: learning rate</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">n_epochs</span></code>: number of training epochs</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">w_rec_g</span></code>: the loss weight for reconstructing original graph structure. If the target dataset contains protein modality, we recommend a low value for w_rec_g (e.g., 0); if it contains epigenome modality, we recommend a high value for w_rec_g (e.g., 1).</p></li>
</ul>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">model</span> <span class="o">=</span> <span class="n">SpaMosaic</span><span class="p">(</span>
    <span class="n">modBatch_dict</span><span class="o">=</span><span class="n">input_dict</span><span class="p">,</span> <span class="n">input_key</span><span class="o">=</span><span class="n">input_key</span><span class="p">,</span>
    <span class="n">batch_key</span><span class="o">=</span><span class="s1">&#39;src&#39;</span><span class="p">,</span> <span class="n">intra_knns</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">inter_knn_base</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">w_g</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span>
    <span class="n">seed</span><span class="o">=</span><span class="mi">1234</span><span class="p">,</span>
    <span class="n">device</span><span class="o">=</span><span class="s1">&#39;cuda:0&#39;</span>
<span class="p">)</span>

<span class="n">model</span><span class="o">.</span><span class="n">train</span><span class="p">(</span><span class="n">net</span><span class="o">=</span><span class="s1">&#39;wlgcn&#39;</span><span class="p">,</span> <span class="n">lr</span><span class="o">=</span><span class="mf">0.001</span><span class="p">,</span> <span class="n">n_epochs</span><span class="o">=</span><span class="mi">60</span><span class="p">,</span> <span class="n">w_rec_g</span><span class="o">=</span><span class="mf">1.</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
batch0: [&#39;rna&#39;]
batch1: [&#39;rna&#39;]
batch2: [&#39;rna&#39;]
------Calculating spatial graph...
The graph contains 23720 edges, 2372 cells.
10.0000 neighbors per cell on average.
------Calculating spatial graph...
The graph contains 24970 edges, 2497 cells.
10.0000 neighbors per cell on average.
------Calculating spatial graph...
The graph contains 23860 edges, 2386 cells.
10.0000 neighbors per cell on average.
Searching MNN within rna
Finding MNN between (s1, s2) using KNN (10, 10)
Finding MNN between (s1, s3) using KNN (10, 10)
Finding MNN between (s2, s3) using KNN (10, 10)
Number of mnn pairs for rna:15914
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|█████████████████████████████████████████████████████████████████████████████████████| 60/60 [00:04&lt;00:00, 12.83it/s]
</pre></div></div>
</div>
<p>Tracking the loss history</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">model</span><span class="o">.</span><span class="n">loss_rec</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[&lt;matplotlib.lines.Line2D at 0x7f516c7541f0&gt;]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/tutorials_integration_horizontal_MB_RNA_horizontal_15_1.png" src="../../../_images/tutorials_integration_horizontal_MB_RNA_horizontal_15_1.png" />
</div>
</div>
</section>
<section id="inference">
<h2>inference<a class="headerlink" href="#inference" title="Permalink to this heading"></a></h2>
<p>After training, SpaMosaic can infer the modality-specific embedding for each section. These embeddings will be directly saved in original <code class="docutils literal notranslate"><span class="pre">anndata</span></code> objects. For example, the RNA-specific embeddings can be accessed by <code class="docutils literal notranslate"><span class="pre">ad1_rna.obsm['emb']</span></code>, <code class="docutils literal notranslate"><span class="pre">ad2_rna.obsm['emb']</span></code>, ….</p>
<p>The following <code class="docutils literal notranslate"><span class="pre">infer_emb</span></code> function will return a new list of <code class="docutils literal notranslate"><span class="pre">anndata</span></code> objects, which save the final/merged embeddings for spatial clustering. The final embeddings can be accessed by <code class="docutils literal notranslate"><span class="pre">ad_embs[0].obsm['merged_emb']</span></code>, <code class="docutils literal notranslate"><span class="pre">ad_embs[1].obsm['merged_emb']</span></code>, …</p>
<p>Since we have only modality in horizontal integration, the final/merged embeddings are exactly the same as the RNA-specific embeddings.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ad_embs</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">infer_emb</span><span class="p">(</span><span class="n">input_dict</span><span class="p">,</span> <span class="n">emb_key</span><span class="o">=</span><span class="s1">&#39;emb&#39;</span><span class="p">,</span> <span class="n">final_latent_key</span><span class="o">=</span><span class="s1">&#39;merged_emb&#39;</span><span class="p">)</span>
<span class="n">ad_mosaic</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">ad_embs</span><span class="p">)</span>
<span class="n">ad_mosaic</span> <span class="o">=</span> <span class="n">utls</span><span class="o">.</span><span class="n">get_umap</span><span class="p">(</span><span class="n">ad_mosaic</span><span class="p">,</span> <span class="n">use_reps</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;merged_emb&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/yanxh/anaconda3/envs/spamosaic-env/lib/python3.8/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
</pre></div></div>
</div>
<p>Visualize the final embeddings using UMAP plots, spot are colored by batch labels</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">utls</span><span class="o">.</span><span class="n">plot_basis</span><span class="p">(</span><span class="n">ad_mosaic</span><span class="p">,</span> <span class="n">basis</span><span class="o">=</span><span class="s1">&#39;merged_emb_umap&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;src&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/tutorials_integration_horizontal_MB_RNA_horizontal_19_0.png" src="../../../_images/tutorials_integration_horizontal_MB_RNA_horizontal_19_0.png" />
</div>
</div>
</section>
<section id="clustering">
<h2>clustering<a class="headerlink" href="#clustering" title="Permalink to this heading"></a></h2>
<p>Parameters for the following <code class="docutils literal notranslate"><span class="pre">utls.clustering</span></code> function:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">n_cluster</span></code>: expected number of clusters</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">used_obsm</span></code>: the target key in the <code class="docutils literal notranslate"><span class="pre">ad_mosaic.obsm</span></code> that holds the embeddings for clustering input</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">algo</span></code>: which clustering algorithm to use, <code class="docutils literal notranslate"><span class="pre">mclust</span></code> or <code class="docutils literal notranslate"><span class="pre">kmeans</span></code>. Sometimes <code class="docutils literal notranslate"><span class="pre">mclust</span></code> can fail to output clustering, it will automatically output the <code class="docutils literal notranslate"><span class="pre">kmeans</span></code> outputs.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">key</span></code>: the key in the row metadata that will hold the output clustering labels.</p></li>
</ul>
<p>Generally, SpaMosaic works better with the mclust algorithm in spatial domain identification task. The final/merged embeddings are used as input for mclust and the clustering labels can be accessed by <code class="docutils literal notranslate"><span class="pre">ad_mosaic.obs['mclust']</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">utls</span><span class="o">.</span><span class="n">clustering</span><span class="p">(</span><span class="n">ad_mosaic</span><span class="p">,</span> <span class="n">n_cluster</span><span class="o">=</span><span class="mi">7</span><span class="p">,</span> <span class="n">used_obsm</span><span class="o">=</span><span class="s1">&#39;merged_emb&#39;</span><span class="p">,</span> <span class="n">algo</span><span class="o">=</span><span class="s1">&#39;mclust&#39;</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="s1">&#39;mclust&#39;</span><span class="p">)</span>
<span class="n">utls</span><span class="o">.</span><span class="n">split_adata_ob</span><span class="p">(</span><span class="n">ad_embs</span><span class="p">,</span> <span class="n">ad_mosaic</span><span class="p">,</span> <span class="s1">&#39;obs&#39;</span><span class="p">,</span> <span class="s1">&#39;mclust&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
R[write to console]:                    __           __
   ____ ___  _____/ /_  _______/ /_
  / __ `__ \/ ___/ / / / / ___/ __/
 / / / / / / /__/ / /_/ (__  ) /_
/_/ /_/ /_/\___/_/\__,_/____/\__/   version 6.0.0
Type &#39;citation(&#34;mclust&#34;)&#39; for citing this R package in publications.

</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
fitting ...
  |======================================================================| 100%
</pre></div></div>
</div>
<p>UMAP plots with batch label and mclust label</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">utls</span><span class="o">.</span><span class="n">plot_basis</span><span class="p">(</span><span class="n">ad_mosaic</span><span class="p">,</span> <span class="n">basis</span><span class="o">=</span><span class="s1">&#39;merged_emb_umap&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;src&#39;</span><span class="p">,</span> <span class="s1">&#39;mclust&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/tutorials_integration_horizontal_MB_RNA_horizontal_24_0.png" src="../../../_images/tutorials_integration_horizontal_MB_RNA_horizontal_24_0.png" />
</div>
</div>
<p>Spatial plots of clustering label for individual sections</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sizes</span> <span class="o">=</span> <span class="p">[</span><span class="mi">100</span><span class="p">,</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">100</span><span class="p">]</span>
<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">ad</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">ad_embs</span><span class="p">):</span>
    <span class="n">utls</span><span class="o">.</span><span class="n">plot_basis</span><span class="p">(</span><span class="n">ad</span><span class="p">,</span> <span class="s1">&#39;spatial&#39;</span><span class="p">,</span> <span class="s1">&#39;mclust&#39;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="n">sizes</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/tutorials_integration_horizontal_MB_RNA_horizontal_26_0.png" src="../../../_images/tutorials_integration_horizontal_MB_RNA_horizontal_26_0.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/tutorials_integration_horizontal_MB_RNA_horizontal_26_1.png" src="../../../_images/tutorials_integration_horizontal_MB_RNA_horizontal_26_1.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/tutorials_integration_horizontal_MB_RNA_horizontal_26_2.png" src="../../../_images/tutorials_integration_horizontal_MB_RNA_horizontal_26_2.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="index.html" class="btn btn-neutral float-left" title="Horizontal Integration" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Tonsil_RNA_horizontal.html" class="btn btn-neutral float-right" title="Integrate three tonsil sections in the RNA modality" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright .</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>